File: frontend/lib/litigation-approvals-compliance/matters.ts
/**
 * Matter Master helpers — single canonical matter namespace.
 */

import type {
  LitigationApprovalsCompliancePayload,
  MatterCategory,
  MatterRecord,
} from '@/lib/schemas/litigation-approvals-compliance';

const CRIMINAL_MATTER_CATEGORIES: ReadonlySet<MatterCategory> = new Set([
  'criminal',
  'economic-offence',
]);

const TAX_MATTER_CATEGORIES: ReadonlySet<MatterCategory> = new Set(['tax']);

export function getMatters(payload: LitigationApprovalsCompliancePayload): MatterRecord[] {
  return payload.litigationAndProceedingsMaster.matters;
}

export function getMatterById(
  payload: LitigationApprovalsCompliancePayload,
  matterId: string,
): MatterRecord | undefined {
  if (!matterId) return undefined;
  return getMatters(payload).find((matter) => matter.matterId === matterId);
}

export function formatMatterLabel(matter: MatterRecord | undefined, fallbackId = ''): string {
  if (!matter) {
    return fallbackId ? `Unknown matter (${fallbackId.slice(0, 8)})` : 'Unknown matter';
  }

  const title = matter.identity.matterTitle.trim() || matter.identity.internalShortName.trim();
  const reference = matter.identity.caseReferenceNumber.trim();
  const category = matter.identity.category.replaceAll('-', ' ');

  const parts = [title, reference, category].filter(Boolean);
  return parts.length > 0 ? parts.join(' — ') : matter.matterId.slice(0, 8);
}

export function isTaxMatter(matter: MatterRecord | undefined): boolean {
  if (!matter?.identity.category) return false;
  return TAX_MATTER_CATEGORIES.has(matter.identity.category);
}

export function isCriminalMatter(matter: MatterRecord | undefined): boolean {
  if (!matter?.identity.category) return false;
  return CRIMINAL_MATTER_CATEGORIES.has(matter.identity.category);
}

File: backend/app/modules/litigation_approvals_compliance/matters.py
"""Matter Master helpers — single canonical matter namespace."""

from __future__ import annotations

from typing import Any

CRIMINAL_MATTER_CATEGORIES = frozenset({"criminal", "economic-offence"})
TAX_MATTER_CATEGORIES = frozenset({"tax"})


def get_matters(payload: dict[str, Any]) -> list[dict[str, Any]]:
    section = payload.get("litigationAndProceedingsMaster") or {}
    return [m for m in (section.get("matters") or []) if isinstance(m, dict)]


def matter_ids(payload: dict[str, Any]) -> set[str]:
    return {str(m.get("matterId")) for m in get_matters(payload) if m.get("matterId")}


def get_matter_by_id(payload: dict[str, Any], matter_id: str) -> dict[str, Any] | None:
    if not matter_id:
        return None
    for matter in get_matters(payload):
        if matter.get("matterId") == matter_id:
            return matter
    return None


def format_matter_label(matter: dict[str, Any] | None, fallback_id: str = "") -> str:
    if not matter:
        if fallback_id:
            return f"Unknown matter ({fallback_id[:8]})"
        return "Unknown matter"

    identity = matter.get("identity") or {}
    title = str(identity.get("matterTitle") or "").strip()
    short_name = str(identity.get("internalShortName") or "").strip()
    reference = str(identity.get("caseReferenceNumber") or "").strip()
    category = str(identity.get("category") or "").replace("-", " ")

    parts = [part for part in (title or short_name, reference, category) if part]
    if parts:
        return " — ".join(parts)
    return str(matter.get("matterId") or fallback_id)[:8]


def is_tax_matter(matter: dict[str, Any] | None) -> bool:
    if not matter:
        return False
    category = str((matter.get("identity") or {}).get("category") or "")
    return category in TAX_MATTER_CATEGORIES


def is_criminal_matter(matter: dict[str, Any] | None) -> bool:
    if not matter:
        return False
    category = str((matter.get("identity") or {}).get("category") or "")
    return category in CRIMINAL_MATTER_CATEGORIES
